Applications/Spire/Source/SpireTester/TableModelTester.cpp

```diff
@@ -1,250 +1,159 @@
 #include <boost/variant/get.hpp>
 #include <doctest/doctest.h>
-#include "Spire/Ui/ArrayTableModel.hpp"
+#include "Spire/Ui/TableModel.hpp"
 
 using namespace boost;
-using namespace boost::signals2;
 using namespace Spire;
 
 TEST_SUITE("TableModel") {
   TEST_CASE("visit_all_operations") {
-    auto model = ArrayTableModel();
-    auto index = 0;
-    auto connection = scoped_connection(model.connect_operation_signal(
-      [&] (const TableModel::Operation& operation) {
-        model.visit(operation,
-          [&] (const TableModel::AddOperation& add_operation) {
-            REQUIRE(add_operation.m_index == index);
-          },
-          [&] (const TableModel::MoveOperation& move_operation) {
-            REQUIRE(move_operation.m_source == 0);
-            REQUIRE(move_operation.m_destination == 1);
-          },
-          [&] (const TableModel::RemoveOperation& remove_operation) {
-            REQUIRE(remove_operation.m_index == index);
-          },
-          [&] (const TableModel::UpdateOperation& update_operation) {
-            REQUIRE(update_operation.m_row == 0);
-            REQUIRE(update_operation.m_column == 0);
-          });
-      }));
-    index = 0;
-    REQUIRE_NOTHROW(model.push({1, 2}));
-    index = 1;
-    REQUIRE_NOTHROW(model.insert({3, 4}, index));
-    REQUIRE_NOTHROW(model.move(0, 1));
-    REQUIRE_NOTHROW(model.set(0, 0, 0));
-    index = 1;
-    REQUIRE_NOTHROW(model.remove(index));
-    index = 0;
-    REQUIRE_NOTHROW(model.remove(index));
+    auto visitor = [&] (const TableModel::Operation& operation) {
+      visit(operation,
+        [&] (const TableModel::AddOperation& add_operation) {
+          REQUIRE(add_operation.m_index == 0);
+        },
+        [&] (const TableModel::MoveOperation& move_operation) {
+          REQUIRE(move_operation.m_source == 0);
+          REQUIRE(move_operation.m_destination == 1);
+        },
+        [&] (const TableModel::RemoveOperation& remove_operation) {
+          REQUIRE(remove_operation.m_index == 1);
+        },
+        [&] (const TableModel::UpdateOperation& update_operation) {
+          REQUIRE(update_operation.m_row == 0);
+          REQUIRE(update_operation.m_column == 0);
+        });
+    };
+    auto add_operation = TableModel::AddOperation{0};
+    visitor(add_operation);
```

Comment:
These tests don't guarantee correctness. For example this unit test would pass if ```visit``` had the following implementation:

    template<typename... F>
    void visit(const TableModel::Operation&, F&&...) {}

One strategy to ensure unit tests work properly is to write a few unit tests before working on the implementation and verifying that they all fail.

Reply:
I think the implementation that you mentioned should be included in the following implementation.
  template<typename F, typename... G>
  void visit(const TableModel::Operation& operation, F&& f, G&&... g) {}

In the unit tests, visiting all operations, visiting partial operations and visiting the default handler have been included, so the two implementations below have been tested:
 template<typename F>
  void visit(const TableModel::Operation& operation, F&& f){}
and
template<typename F, typename... G>
  void visit(const TableModel::Operation& operation, F&& f, G&&... g) {}

Reply:
My point is that your unit tests are not testing anything. If you submitted a pull request with the following implementation:

    template<typename... F>
    void visit(const TableModel::Operation&, F&&...) {}

Almost all of your unit tests would pass. In order for a unit test to be valid, they must all fail given the above implementation.